docs/TICCL-stats.1
.TH TICCL-stats 1 "2020 mar 09"

.SH NAME
TICCL-stats - gather n-gram statistics from textfiles (Use the more developed FoLiA-stats for FoLiA XML input!)

.SH SYNOPSIS
TICCL-stats [options] FILE

TICCL-stats [options] DIR

.SH DESCRIPTION

When a DIR is provided,
.B TICCL-stats
will process all files in DIR and store its results in the current
directory in files called DIR.wordfreqlist.tsv

When a FILE is provided,
.B TICCL-stats
will process that file and store its results in the directory where FILE is
found

The output format will be 2 or 4 <tab> separated columns (depending on the
.B -p
option)

First column:
the 'word' tag at hand.

Second column:
the corpus frequency of the 'word'

when -p is provided:

Third column:
the accumulated frequencies of all entries up to and including the current one.

Fourth column:
the relative presence of the entry: which percentage of the corpus does the
entry belong to?

.SH OPTIONS
.B --clip
number
.RS
clipping factor. When an item's frequency is lower than 'number', it will not be stored.
.RE

.B -p
.RS
Also output accumulated counts and percentages
.RE

.B -n
.RS
read the input line by line.
.RE

.B -o
outprefix
.RS
use outprefix for all output files.
.RE

.B --lower
.RS
Lowercase all words.
.RE

.B --separator
sep
.RS
Define a separator character to connect ngrams. Default is an underscore.
.RE

.B --underscore
.RS
Backward compatibility. Preferably use --separator=_
.RE

.B --ngram
count
.RS
Construct all n-grams of length 'count'
.RE

.B --hemp
file
.RS
Create a historical emphasis file. This separately collects words consisting of, i.e. originally printed or written as singe space-separated letters.
.RE

.B -t
or
.B --threads
num_threads
.RS
use 'num_threads' threads to run on. You may us --threads="max" to use as many
threads as possible. This will allocate 2 processors less than given by the
$OMP_NUM_THREADS environment variable, leaving some processor power for other
purposes.
.RE

.B -e
expr
.RS
when searching for files,
.B
TICCL-stats
will only consider files that match with the expression 'expr' which may contain wildcards. The 'expr' is only matched against the file part. Not against paths.
.RE

.B -R
.RS
when a DIR is provided,
.B TICCL-stats
will recurse through this DIR and its subdirs to find files.
.RE

.B -X
.RS
all input is assumed to be XML, and ALL TEXT nodes are used for extracting word strings from
.RE

.B -V
or
.B --version
.RS
Show VERSION
.RE

.B -h
or
.B --help
.RS
Show usage() information
.RE


.SH BUGS
possibly

.SH AUTHORS
Ko van der Sloot lamasoftware@science.ru.nl

Martin Reynaert: reynaert@uvt.nl


NEWS
ticcltools 0.6 2018-06-05
[Ko vander Sloot]
Intermediate release, with a lot of new code to handle N-grams
Also a lot of refactoring is done, for more clear and maintainable code.
This is work in progress still.

* TICCL-unk:
  - more extensive acronym detection
  - fixed artifreq problems in 'clean' punctuated words
  - added filters for 'unwanted' characters
  - added a ligature filter to convert evil ligatures
  - normalize all hyphens to a 'normal' one (-)
  - use a better definition of punctuation (unicode character class is not
    good enough to decide)

* TICCL-lexstat:
  - the 'separator' symbol should get freq=0, so it isn't counted
  - the clip value is added to the output filename

* TICCL-indexer:
  - indexer and indexerNT now produce the same output, using different
    strategies when a --foci files is used.

* TICCL-LDcalc:
  major overhaul for n-grams
  - added a ngram point column to the output (so NOT backward compatible!)
  - produce a '.short' list for short word corrections
  - produce a '.ambi' file with a list of n-grams related to short words
  - prune a lot of ngrams from the output

* TICCL-rank:
 - output is sorted now
 - honor the ngram-points from the new LDcalc. (so NOT backward compatible!)

* TICCL-chain: new module to chain ranked files

* TICCL-lexclean:
  -added a -x option for 'inverse' alphabet

* TICCL-anahash:
  - added a --list option to produce a list of words and anagram values

[Maarten van Gompel]
* added metadata file: codemeta.json

ticcltools 0.5 2018-02-19
[Ko van der Sloot]
* updated configuration. also for Mac OSX
* use of more ticcutils stuff: diacritics filter
* added a TICCL-mergelex program
* the OMP_THREAD_LIMIT environment variable was ignored sometimes
* TICCL-unk:
   - fixed a problem in artifreq handling
   - changed acronym detection (work in  progress)
   - added -o option
  TICCL-lexstat:
   - added TTR output
   - added -o option
  TICCL-indexer
  - now also handles --foci file. with some speed-up
  - added a -t option
  TICCL-LDcalc:
  - be less picky on a few wrong lines in the data
* added some tests
* when libroaring is installed we built roaring versions of some modules (experimental)
* updated man pages

ticcltools 0.4 2017-04-04
[Ko van der Sloot]
* first official release.
  - added functions to test on Word2Vec datafiles
  - refactoring and modernizing stuff all around

ticcltools 0.3 2016-01-21
* upload ready
ticcltools 0.2 2016-01-14
[Ko van der Sloot]
* repository moved to GitHub
* Travis support
* first 'distributable' version
* added TICCL-stats program
* added W2V-near and W2V-dist programs bases on new libticcl lib


[0.1] Ko vd Sloot
 started autoconfiscating


src/TICCL-stats.cxx
/*
  Copyright (c) 2006 - 2020
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of ticcltools

  foliatools is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  foliatools is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/frog/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl

*/

#include <cmath>
#include <string>
#include <map>
#include <vector>
#include <iostream>
#include <fstream>

#include "ticcutils/CommandLine.h"
#include "ticcutils/FileUtils.h"
#include "ticcutils/StringOps.h"
#include "ticcutils/XMLtools.h"
#include "ticcutils/Unicode.h"

#include "config.h"
#ifdef HAVE_OPENMP
#include "omp.h"
#endif

using namespace	std;
using namespace	TiCC;

bool verbose = false;

void create_wf_list( const map<string, unsigned int>& wc,
		     const string& filename, unsigned int totalIn,
		     unsigned int clip,
		     bool doperc ){
  unsigned int total = totalIn;
  ofstream os( filename );
  if ( !os ){
    cerr << "failed to create outputfile '" << filename << "'" << endl;
    exit(EXIT_FAILURE);
  }
  map<unsigned int, set<string> > wf;
  for( const auto& cit : wc ){
    if ( cit.second <= clip ){
      total -= cit.second;
    }
    else {
      wf[cit.second].insert( cit.first );
    }
  }
  unsigned int sum=0;
  unsigned int types=0;
  map<unsigned int, set<string> >::const_reverse_iterator wit = wf.rbegin();
  while ( wit != wf.rend() ){
    for( const auto& sit : wit->second ){
      sum += wit->first;
      os << sit << "\t" << wit->first;
      if ( doperc ){
	os << "\t" << sum << "\t" << 100 * double(sum)/total;
      }
      os << endl;
      ++types;
    }
    ++wit;
  }
#pragma omp critical
  {
    cout << "created WordFreq list '" << filename << "'" << endl
	 << "with " << total << " tokens and " << types
	 << " types. TTR= " << (double)types/total
	 << ", the angle is " << atan((double)types/total)*180/M_PI
	 << " degrees";
    if ( clip > 0 ){
      cout << "(" << totalIn - total << " of the original " << totalIn
	   << " words were clipped.)";
    }
    cout << endl;
  }
}

static void error_sink(void *mydata, xmlError *error ){
  int *cnt = (int*)mydata;
  if ( *cnt == 0 ){
    cerr << "\nXML-error: " << error->message << endl;
  }
  (*cnt)++;
}

bool is_emph( const string& data ){
  return (data.size() < 2) && isalnum(data[0]);
}

size_t tel( const xmlNode *node, bool lowercase,
	    size_t ngram, const string& sep,
	    map<string, unsigned int>& wc,
	    set<string>& emps ){
  vector<string> buffer(ngram);
  size_t cnt = 0;
  size_t buf_cnt = 0;
  bool in_emph = false;
  string emph_start;
  string emph_word;
  xmlNode *pnt = node->children;
  while ( pnt ){
    //    cerr << "bekijk label: " << (char*)pnt->name << endl;
    cnt += tel( pnt, lowercase, ngram, sep, wc, emps );
    if ( pnt->type == XML_TEXT_NODE ){
      string line  = (char*)( pnt->content );
      //      cerr << "text: " << line << endl;
      vector<string> v;
      TiCC::split( line, v );
      for ( const auto& word : v ){
	string wrd = word;
	if ( lowercase ){
	  wrd = TiCC::utf8_lowercase( word );
	}
	if ( is_emph( wrd ) ){
	  if ( in_emph ){
	    emph_word += "_" + wrd;
	  }
	  else {
	    emph_start = wrd;
	    in_emph = true;
	  }
	}
	else {
	  if ( in_emph && !emph_word.empty() ){
	    emps.insert( emph_start + emph_word );
	  }
	  in_emph = false;
	  emph_start.clear();
	  emph_word.clear();
	}
	buffer[buf_cnt++] = wrd;
	if ( buf_cnt == ngram ){
	  buf_cnt = ngram-1;
	  string gram;
	  for( size_t i=0; i < ngram -1; ++i ){
	    gram += buffer[i] + sep;
	    buffer[i] = buffer[i+1];
	  }
	  gram+= buffer[ngram-1];
#pragma omp critical
	  {
	    ++wc[gram];
	  }
	  ++cnt;
	}
      }
    }
    pnt = pnt->next;
  }
  return cnt;
}

size_t word_xml_inventory( const string& docName,
			   bool lowercase,
			   size_t ngram,
			   const string& sep,
			   map<string,unsigned int>& wc,
			   set<string>& emps ){
  xmlDoc *d = 0;
  int cnt = 0;
  xmlSetStructuredErrorFunc( &cnt, (xmlStructuredErrorFunc)error_sink );
  d = xmlReadFile( docName.c_str(), 0, XML_PARSE_NOBLANKS|XML_PARSE_HUGE );
  if ( !d || cnt > 0 ){
#pragma omp critical
    {
      cerr << "failed to load document '" << docName << "'" << endl;
    }
    return 0;
  }
  xmlNode *root = xmlDocGetRootElement( d );
  size_t wordTotal = tel( root, lowercase, ngram, sep, wc, emps );
  xmlFree( d );
  return wordTotal;
}

size_t word_inventory( const string& docName,
		       bool lowercase,
		       size_t ngram,
		       const string& sep,
		       map<string,unsigned int>& wc,
		       set<string>& emps,
		       bool dolines ){
  vector<string> buffer(ngram);
  size_t buf_cnt = 0;
  size_t wordTotal = 0;
  ifstream is( docName );
  string line;
  bool in_emph = false;
  string emph_start;
  string emph_word;
  while ( getline( is, line ) ){
    if ( dolines ){
      buffer.clear();
      buf_cnt = 0;
    }
    vector<string> v;
    TiCC::split( line, v );
    for ( const auto& word : v ){
      string wrd = word;
      if ( lowercase ){
	wrd = TiCC::utf8_lowercase( word );
      }
      if ( is_emph( wrd ) ){
	if ( in_emph ){
	  emph_word += "_" + wrd;
	}
	else {
	  emph_start = wrd;
	  in_emph = true;
	}
      }
      else {
	if ( in_emph && !emph_word.empty() ){
	  emps.insert( emph_start + emph_word );
	}
	in_emph = false;
	emph_start.clear();
	emph_word.clear();
      }
      buffer[buf_cnt++] = wrd;
      if ( buf_cnt == ngram ){
	buf_cnt = ngram-1;
	string gram;
	for( size_t i=0; i < ngram -1; ++i ){
	  gram += buffer[i] + sep;
	  buffer[i] = buffer[i+1];
	}
	gram += buffer[ngram-1];
#pragma omp critical
	{
	  ++wc[gram];
	}
	++wordTotal;
      }
    }
  }
  return wordTotal;
}


void usage( const string& name ){
  cerr << "Usage: " << name << " [options] file/dir" << endl;
  cerr << "\t TICCL-stats will produce ngram statistics for a file, " << endl;
  cerr << "\t or a whole directory of files " << endl;
  cerr << "\t The output will be a 2 columned tab separated file, extension: *tsv " << endl;
  cerr << "\t--clip\t clipping factor. " << endl;
  cerr << "\t\t\t(entries with frequency <= this factor will be ignored). " << endl;
  cerr << "\t-p\t output percentages too. " << endl;
  cerr << "\t--lower\t Lowercase all words" << endl;
  cerr << "\t--ngram size\t create an ngram for 'size' ngrams (default 1-gram)" << endl;
  cerr << "\t--separator='sep' 	connect all n-grams with 'sep' (default is an underscore)" << endl;
  cerr << "\t--underscore\t Same as --separator='_'" << endl;
  cerr << "\t--hemp=<file>. Create a historical emphasis file. " << endl;
  cerr << "\t-t <threads> or --threads <threads> Number of threads to run on." << endl;
  cerr << "\t\t\t If 'threads' has the value \"max\", the number of threads is set to a" << endl;
  cerr << "\t\t\t reasonable value. (OMP_NUM_TREADS - 2)" << endl;
  cerr << "\t-n\t newlines delimit the input." << endl;
  cerr << "\t-v\t very verbose output." << endl;
  cerr << "\t-e expr\t specify the expression all input files should match with." << endl;
  cerr << "\t-o\t name of the output file(s) prefix." << endl;
  cerr << "\t-X\t the inputfiles are assumed to be XML. (all TEXT nodes are used)" << endl;
  cerr << "\t-R\t search the dirs recursively (when appropriate)." << endl;
  cerr << "\t-V or --version\t show version " << endl;
  cerr << "\t-h or --help \t this message." << endl;
}

int main( int argc, char *argv[] ){
  CL_Options opts( "hnVvpe:t:o:RX", "clip:,lower,ngram:,underscore,separator:,hemp:,threads:" );
  try {
    opts.init(argc,argv);
  }
  catch( const OptionError& e ){
    cerr << e.what() << endl;
    usage(argv[0]);
    exit( EXIT_FAILURE );
  }
  string progname = opts.prog_name();
  if ( argc < 2 ){
    usage( progname );
    exit(EXIT_FAILURE);
  }
  int clip = 0;
  string expression;
  string outputPrefix;
  if ( opts.extract('V' ) || opts.extract("version") ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract('h' ) || opts.extract("help") ){
    usage(progname);
    exit(EXIT_SUCCESS);
  }
  verbose = opts.extract( 'v' );
  bool dolines = opts.extract( 'n' );
  bool doXML = opts.extract( 'X' );
  if ( doXML && dolines ){
    cerr << "options -X and -n conflict!" << endl;
  }
  bool dopercentage = opts.extract('p');
  bool lowercase = opts.extract("lower");
  bool recursiveDirs = opts.extract( 'R' );
  string sep = "_";
  opts.extract( "separator", sep );
  if ( opts.extract( "underscore" ) ){
    if ( sep != "_" ){
      cerr << "--separator and --underscore conflict!" << endl;
      exit(EXIT_FAILURE);
    }
    sep = "_";
  }
  string hempName;
  opts.extract("hemp", hempName );
  if ( !opts.extract( 'o', outputPrefix ) ){
    cerr << "an output filename prefix is required. (-o option) " << endl;
    exit(EXIT_FAILURE);
  }
  string value;
  if ( opts.extract("clip", value ) ){
    if ( !stringTo(value, clip ) ){
      cerr << "illegal value for --clip (" << value << ")" << endl;
      exit(EXIT_FAILURE);
    }
  }
  size_t ngram = 1;
  if ( opts.extract("ngram", value ) ){
    if ( !stringTo(value, ngram )
	 || (ngram < 1) ){
      cerr << "illegal value for --ngram (" << value << ")" << endl;
      exit(EXIT_FAILURE);
    }
  }
  value = "1";
  if ( !opts.extract( 't', value ) ){
    opts.extract( "threads", value );
  }
#ifdef HAVE_OPENMP
  int numThreads=1;
  if ( TiCC::lowercase(value) == "max" ){
    numThreads = omp_get_max_threads() - 2;
    omp_set_num_threads( numThreads );
    cout << "running on " << numThreads << " threads." << endl;
  }
  else {
    if ( !TiCC::stringTo(value,numThreads) ) {
      cerr << "illegal value for -t (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
    omp_set_num_threads( numThreads );
    cout << "running on " << numThreads << " threads." << endl;
  }
#else
  if ( value != "1" ){
    cerr << "unable to set number of threads!.\nNo OpenMP support available!"
	 <<endl;
    exit(EXIT_FAILURE);
  }
#endif
  opts.extract('e', expression );
  if ( !opts.empty() ){
    cerr << "unsupported options : " << opts.toString() << endl;
    usage(progname);
    exit(EXIT_FAILURE);
  }

  vector<string> massOpts = opts.getMassOpts();
  if ( massOpts.empty() ){
    cerr << "no file or dir specified!" << endl;
    exit(EXIT_FAILURE);
  }
  string name = massOpts[0];
  vector<string> fileNames = searchFilesMatch( name, expression, recursiveDirs );
  size_t toDo = fileNames.size();
  if ( toDo == 0 ){
    cerr << "no matching files found" << endl;
    exit(EXIT_SUCCESS);
  }

  string::size_type pos = outputPrefix.find( "." );
  if ( pos != string::npos && pos == outputPrefix.length()-1 ){
    // outputname ends with a .
    outputPrefix = outputPrefix.substr(0,pos);
  }
  pos = outputPrefix.find( "/" );
  if ( pos != string::npos && pos == outputPrefix.length()-1 ){
    // outputname ends with a /
    outputPrefix += "ticclstats";
  }

  if ( toDo > 1 ){
    cout << "start processing of " << toDo << " files " << endl;
  }
  map<string,unsigned int> wc;
  unsigned int wordTotal =0;

  set<string> hemp;
#pragma omp parallel for shared(fileNames,wordTotal,wc,hemp)
  for ( size_t fn=0; fn < fileNames.size(); ++fn ){
    string docName = fileNames[fn];
    unsigned int word_count =  0;
    if ( doXML ){
      word_count = word_xml_inventory( docName, lowercase, ngram, sep, wc, hemp );
    }
    else {
      word_count = word_inventory( docName, lowercase, ngram, sep, wc, hemp, dolines );
    }
    wordTotal += word_count;
#pragma omp critical
    {
      cout << "Processed :" << docName << " with " << word_count << " words,"
	   << " still " << --toDo << " files to go." << endl;
    }
  }
  if ( toDo > 1 ){
    cout << "done processsing directory '" << name << "' in total "
	 << wordTotal << " words were found." << endl;
  }
  cout << "start calculating the results" << endl;
  if ( !hempName.empty() ){
    ofstream out( hempName );
    if ( out ){
      for( auto const& it : hemp ){
	out << it << endl;
      }
      cout << "historical emphasis stored in: " << hempName << endl;
    }
    else {
      cerr << "unable to create historical emphasis file: " << hempName << endl;
    }
  }
  string filename = outputPrefix + ".wordfreqlist";
  string path = dirname( filename ) + "/";
  if ( !TiCC::createPath( path ) ){
    cerr << "unable to create a path: " << path << endl;
    exit(EXIT_FAILURE);
  }
  string ng = toString(ngram);
  filename += "." + ng + ".tsv";
  create_wf_list( wc, filename, wordTotal, clip, dopercentage );
  exit( EXIT_SUCCESS );
}
